cmd/cloudFoundryDeploy.go

```diff
@@ -0,0 +1,855 @@
+				}
+			} else {
+				return map[string]interface{}{}, fmt.Errorf("Cannot cast key '%v' to string", key)
+			}
+		}
+	}
+
+	return out, err
+}
+func getVarOptions(vars []string) ([]string, error) {
+
+	varsMap, err := toParameterMap(vars)
+	if err != nil {
+		return []string{}, err
+	}
+
+	varsResult := []string{}
+
+	for _, key := range varsMap.Keys() {
+		val, _ := varsMap.Get(key)
+		if v, ok := val.(string); ok {
+			varsResult = append(varsResult, "--var", fmt.Sprintf("%s=%s", key, quoteAndBashEscape(v)))
+		} else {
+			return []string{}, fmt.Errorf("Cannot cast '%v' to string", val)
+		}
+	}
+	return varsResult, nil
+}
+
+func getVarFileOptions(varFiles []string) ([]string, error) {
+
+	varFilesResult := []string{}
+
+	for _, varFile := range varFiles {
+		fExists, err := fileUtils.FileExists(varFile)
+		if err != nil {
+			return []string{}, err
+		}
+
+		if !fExists {
+			log.Entry().Warningf("We skip adding not-existing file '%s' as a vars-file to the cf create-service-push call", varFile)
+			continue
+		}
+
+		varFilesResult = append(varFilesResult, "--vars-file", quoteAndBashEscape(varFile))
+	}
+
+	if len(varFilesResult) > 0 {
+		log.Entry().Infof("We will add the following string to the cf push call: '%s'", strings.Join(varFilesResult, " "))
+	}
+	return varFilesResult, nil
+}
+
+func quoteAndBashEscape(s string) string {
+	escapedSingleQuote := "'\"'\"'"
+	return strings.ReplaceAll(s, "'", escapedSingleQuote)
+}
+
+func checkAndUpdateDeployTypeForNotSupportedManifest(config *cloudFoundryDeployOptions) (string, error) {
+
+	manifestFile := config.Manifest
+	if len(manifestFile) == 0 {
+		manifestFile = "manifest.yml"
+	}
+	var manifestFileExists bool
+	var err error
+
+	if len(manifestFile) > 0 {
+		manifestFileExists, err = fileUtils.FileExists(manifestFile)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if config.DeployType == "blue-green" && manifestFileExists {
+
+		m, _ := _getManifest(manifestFile)
+
+		apps, err := m.GetApplications()
+
+		if err != nil {
+			return "", err
+		}
+		if len(apps) > 1 {
+			return "", fmt.Errorf("Your manifest contains more than one application. For blue green deployments your manifest file may contain only one application")
+		}
+
+		hasNoRouteProperty, err := m.ApplicationHasProperty(0, "no-route")
+		if err != nil {
+			return "", err
+		}
+		if len(apps) == 1 && hasNoRouteProperty {
+
+			const deployTypeStandard = "standard"
+			log.Entry().Warningf("Blue green deployment is not possible for application without route. Using deployment type '%s' instead.", deployTypeStandard)
+			return deployTypeStandard, nil
+		}
+	}
+
+	return config.DeployType, nil
+}
+
+func deployMta(config *cloudFoundryDeployOptions, mtarFilePath string, command command.ExecRunner) error {
+
+	deployCommand := "deploy"
+	deployParams := []string{}
+
+	if len(config.MtaDeployParameters) > 0 {
+		deployParams = append(deployParams, strings.Split(config.MtaDeployParameters, " ")...)
```

Comment:
I am wondering whether mtaDeploayParameters should be of type []string. It is not possible to handle spaces in the parameters, neither here, nor by the caller of the step.

Reply:
I had the same thought, but I hesitated to do this. Reason: backward-compatibiity. We might have old configs where it is a string.

We have two cases where I used []string in the metadata for the step. In these two cases we have also lists in the groovy part.